Variants where called following the GATK best practices workflow: Reads were mapped onto {{ snakemake.config["ref"]["species"] }} build {{ snakemake.config["ref"]["build"] }} with BWA mem, and both optical and PCR duplicates were removed with Picard, followed by base recalibration with GATK. The GATK HaplotypeCaller was used to call variants per-sample, including summarized evidence for non-variant sites (GVCF approach). Then, GATK genotyping was done in a joint way over GVCF files of all samples. {% if snakemake.config["filtering"]["vqsr"] %} Genotyped variants were filtered with the GATK VariantRecalibrator approach. {% else %} Genotyped variants were filtered using hard thresholds. For SNVs, the criterion {{ snakemake.config["filtering"]["hard"]["snvs"] }} was used, for Indels the criterion {{ snakemake.config["filtering"]["hard"]["indels"] }} was used. {% endif %} Finally, SnpEff was used to predict and report variant effects. In addition, quality control was performed with FastQC, Samtools, and Picard and aggregated into an interactive report via MultiQC.