{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4f6c2f1e-029b-432b-a913-511fe50044f6",
   "metadata": {},
   "source": [
    "# Overall Explanation of Generating Simulated Sequence File "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9e1d9a-b3b1-45f5-84a0-04a9b4d97ee4",
   "metadata": {},
   "source": [
    "explanation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74070417-acd8-4fdd-a3ac-6db2ab0673fe",
   "metadata": {},
   "source": [
    "# Load Golden Standard Set From "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8018a6ba-e78e-493b-9df5-9b8b13428272",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b4835c03-9986-4393-811e-2851527fa365",
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "import subprocess\n",
    "import pysam\n",
    "import pyranges \n",
    "\n",
    "\n",
    "\n",
    "def vcf_to_bed(vcf_path, bed_path):\n",
    "    # Open VCF file\n",
    "    vcf_reader = vcf.Reader(open(vcf_path, 'r'))\n",
    "\n",
    "    # Open BED file for writing\n",
    "    with open(bed_path, 'w') as bed_file:\n",
    "        # Iterate through VCF records and write to BED file\n",
    "        for record in vcf_reader:\n",
    "            chrom = record.CHROM\n",
    "            start = record.POS - 1  # Convert to 0-based coordinates for BED format\n",
    "            end = record.POS\n",
    "            info = record.INFO\n",
    "\n",
    "            bed_file.write(f\"{chrom}\\t{start}\\t{end}\\t{info}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bb91bf2e-e946-4b1c-8302-07d3318af711",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_to_bed(\"results/xTea/shortread/mosaic/GIAB/Illumina300x/HG002/Alu/HG002.GRCh38.300x_ALU.vcf\", \"results/xTea/shortread/mosaic/GIAB/Illumina300x/HG002/Alu/HG002.GRCh38.300x_ALU.bed\")\n",
    "vcf_to_bed(\"results/xTea/shortread/mosaic/GIAB/Illumina300x/HG003/Alu/HG003.GRCh38.300x_ALU.vcf\", \"results/xTea/shortread/mosaic/GIAB/Illumina300x/HG003/Alu/HG003.GRCh38.300x_ALU.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "569d621f-f18b-4a22-a114-fb551d06af1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jp394/miniforge3/envs/SMaTH/lib/python3.10/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>20606196</td>\n",
       "      <td>20606197</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 95, 'END': 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>23654636</td>\n",
       "      <td>23654637</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 316, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>26904034</td>\n",
       "      <td>26904035</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 452, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29100312</td>\n",
       "      <td>29100313</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 466, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>37775709</td>\n",
       "      <td>37775710</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 210, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>chr22</td>\n",
       "      <td>46227137</td>\n",
       "      <td>46227138</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 250, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>chrX</td>\n",
       "      <td>255569</td>\n",
       "      <td>255570</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 270, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>chrX</td>\n",
       "      <td>31982954</td>\n",
       "      <td>31982955</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 265, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>chrX</td>\n",
       "      <td>117704900</td>\n",
       "      <td>117704901</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 188, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>chrX</td>\n",
       "      <td>152833245</td>\n",
       "      <td>152833246</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 385, 'END': ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>276 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 20606196  | 20606197  | ...   |\n",
       "| chr1         | 23654636  | 23654637  | ...   |\n",
       "| chr1         | 26904034  | 26904035  | ...   |\n",
       "| chr1         | 29100312  | 29100313  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chrX         | 255569    | 255570    | ...   |\n",
       "| chrX         | 31982954  | 31982955  | ...   |\n",
       "| chrX         | 117704900 | 117704901 | ...   |\n",
       "| chrX         | 152833245 | 152833246 | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 276 rows and 4 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyranges.read_bed(\"results/xTea/shortread/mosaic/GIAB/Illumina300x/HG003/Alu/HG003.GRCh38.300x_ALU.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "80dbc3b7-c0a6-4e92-ac24-54caca8cb99c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jp394/miniforge3/envs/SMaTH/lib/python3.10/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21409316</td>\n",
       "      <td>21409317</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 104, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28994949</td>\n",
       "      <td>28994950</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 286, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29100312</td>\n",
       "      <td>29100313</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 299, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>37775709</td>\n",
       "      <td>37775710</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 208, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>37967092</td>\n",
       "      <td>37967093</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 296, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>chr22</td>\n",
       "      <td>41314923</td>\n",
       "      <td>41314924</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 270, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>chr22</td>\n",
       "      <td>49879715</td>\n",
       "      <td>49879716</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 462, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>chrX</td>\n",
       "      <td>34446213</td>\n",
       "      <td>34446214</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 295, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>chrX</td>\n",
       "      <td>117704886</td>\n",
       "      <td>117704887</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 249, 'END': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56886097</td>\n",
       "      <td>56886098</td>\n",
       "      <td>{'SVTYPE': 'INS:ME:ALU', 'SVLEN': 44, 'END': 5...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>301 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Start     | End       | +1    |\n",
       "| (category)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+-----------+-----------+-------|\n",
       "| chr1         | 21409316  | 21409317  | ...   |\n",
       "| chr1         | 28994949  | 28994950  | ...   |\n",
       "| chr1         | 29100312  | 29100313  | ...   |\n",
       "| chr1         | 37775709  | 37775710  | ...   |\n",
       "| ...          | ...       | ...       | ...   |\n",
       "| chr22        | 49879715  | 49879716  | ...   |\n",
       "| chr22        | 41314923  | 41314924  | ...   |\n",
       "| chr22        | 35631628  | 35631629  | ...   |\n",
       "| chr22        | 23530480  | 23530481  | ...   |\n",
       "| chrX         | 117704886 | 117704887 | ...   |\n",
       "| chrX         | 34446213  | 34446214  | ...   |\n",
       "| chrY         | 56886097  | 56886098  | ...   |\n",
       "+--------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 301 rows and 4 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "1 hidden columns: Name"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyranges.read_bed(\"results/xTea/shortread/mosaic/GIAB/Illumina300x/HG002/Alu/HG002.GRCh38.300x_ALU.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2187776a-b5b5-49a6-be5a-d53838d6d4f4",
   "metadata": {},
   "source": [
    "# Remove Spikes from BAM File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aa891b36-462a-48cf-8d4e-d961634c3e54",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def remove_spikes(input_bam, bed_file, output_bam):\n",
    "    try:\n",
    "        # Open input BAM file\n",
    "        with pysam.AlignmentFile(input_bam, \"rb\") as input_bam_file:\n",
    "            # Open BED file for regions to keep\n",
    "            with open(bed_file, \"r\") as bed:\n",
    "                # Create a list of regions from the BED file\n",
    "                regions = [line.strip().split('\\t') for line in bed]\n",
    "\n",
    "                # Fetch reads overlapping the specified regions\n",
    "                filtered_reads = input_bam_file.fetch(regions=regions)\n",
    "\n",
    "                # Open output BAM file for writing\n",
    "                with pysam.AlignmentFile(output_bam, \"wb\", header=input_bam_file.header) as output_bam_file:\n",
    "                    # Write filtered reads to output BAM file\n",
    "                    for read in filtered_reads:\n",
    "                        output_bam_file.write(read)\n",
    "\n",
    "        print(f\"Spikes removed successfully. Output saved to {output_bam}\")\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error: {e}\")\n",
    "\n",
    "def extract_regions(input_bam, bed_file, output_bam):\n",
    "    # Step 2: Extract regions from the first BAM file\n",
    "    with pysam.AlignmentFile(input_bam, \"rb\") as bam_file:\n",
    "        regions = pysam.TabixFile(bed_file)\n",
    "        extracted_reads = [read for read in bam_file.fetch() if regions.fetch(read.reference_name, read.reference_start, read.reference_end)]\n",
    "\n",
    "    # Write the extracted reads to a new BAM file\n",
    "    with pysam.AlignmentFile(output_bam, \"wb\", header=bam_file.header) as output_bam_file:\n",
    "        for read in extracted_reads:\n",
    "            output_bam_file.write(read)\n",
    "\n",
    "\n",
    "def simulate_insertions(reference_genome, extracted_bam, output_prefix):\n",
    "    try:\n",
    "        # Open reference genome\n",
    "        with pysam.FastaFile(reference_genome) as ref_fasta:\n",
    "            # Open extracted BAM file\n",
    "            with pysam.AlignmentFile(extracted_bam, \"rb\") as input_bam_file:\n",
    "                # Open output BAM file for writing simulated reads\n",
    "                with pysam.AlignmentFile(output_prefix + \".bam\", \"wb\", header=input_bam_file.header) as output_bam_file:\n",
    "                    # Iterate over reads in the input BAM file\n",
    "                    for read in input_bam_file:\n",
    "                        # Simulate insertions (modify read sequence, qualities, etc. as needed)\n",
    "                        simulated_read = simulate_insertion(read, ref_fasta)\n",
    "\n",
    "                        # Write the simulated read to the output BAM file\n",
    "                        output_bam_file.write(simulated_read)\n",
    "\n",
    "        print(f\"Insertions simulated successfully. Output saved to {output_prefix}.bam\")\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error: {e}\")\n",
    "\n",
    "def simulate_insertion(read, ref_fasta):\n",
    "    # Modify the read to simulate an insertion (customize based on your needs)\n",
    "    # For example, you can extend the read sequence, qualities, etc.\n",
    "    simulated_read = read.copy()\n",
    "    simulated_read.query_sequence += \"INSERTED_SEQUENCE\"\n",
    "    simulated_read.query_qualities += [30] * len(\"INSERTED_SEQUENCE\")\n",
    "\n",
    "    # Adjust alignment information if necessary\n",
    "    simulated_read.reference_end += len(\"INSERTED_SEQUENCE\")\n",
    "\n",
    "    return simulated_read\n",
    "\n",
    "\n",
    "def merge_bams(original_bam, simulated_bam, merged_bam):\n",
    "    # Step 4: Merge BAM files\n",
    "    with pysam.AlignmentFile(original_bam, \"rb\") as original_file, \\\n",
    "         pysam.AlignmentFile(simulated_bam + '.sam', \"rb\") as simulated_file, \\\n",
    "         pysam.AlignmentFile(merged_bam, \"wb\", header=original_file.header) as output_bam:\n",
    "        for read in original_file:\n",
    "            output_bam.write(read)\n",
    "        for read in simulated_file:\n",
    "            output_bam.write(read)\n",
    "\n",
    "    # Step 5: Sort and index merged BAM file\n",
    "    pysam.sort('-o', merged_bam.replace('.bam', '_sorted.bam'), merged_bam)\n",
    "    pysam.index(merged_bam.replace('.bam', '_sorted.bam'))\n",
    "\n",
    "# if __name__ == \"__main__\":\n",
    "#     # Specify file paths\n",
    "#     input_bam = \"/path/to/your/original.bam\"\n",
    "#     bed_file = \"/path/to/your/simulated_insertions.bed\"\n",
    "#     output_bam = \"/path/to/your/extracted_reads.bam\"\n",
    "#     reference_genome = \"/path/to/your/reference_genome.fasta\"\n",
    "#     output_prefix = \"/path/to/your/simulated_insertions\"\n",
    "#     merged_bam = \"/path/to/your/final_merged.bam\"\n",
    "\n",
    "#     # Extract regions from the original BAM file\n",
    "#     extract_regions(input_bam, bed_file, output_bam)\n",
    "\n",
    "#     # Simulate insertions\n",
    "#     simulate_insertions(reference_genome, output_bam, output_prefix)\n",
    "\n",
    "#     # Merge the original and simulated BAM files\n",
    "#     merge_bams(input_bam, output_prefix + '.bam', merged_bam)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca923d1c-d312-4400-b805-1975dce92dbe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
